packages/phone-number-privacy/combiner/src/match-making/get-contact-matches.ts

```diff
@@ -36,16 +44,102 @@ export async function handleGetContactMatches(
       return
     }
 
-    const { account, userPhoneNumber, contactPhoneNumbers, hashedPhoneNumber } = request.body
+    const {
+      account,
+      userPhoneNumber,
+      contactPhoneNumbers,
+      hashedPhoneNumber,
+      signedUserPhoneNumber,
+    } = request.body
 
     if (!(await isVerified(account, hashedPhoneNumber, getContractKit(), logger))) {
       respondWithError(response, 403, WarningMessage.UNVERIFIED_USER_ATTEMPT_TO_MATCHMAKE, logger)
       return
     }
+
+    // Verify that signedUserPhoneNumber was signed by the account's DEK
+    let signedUserPhoneNumberHash: string | undefined
+    let shouldStoreRequest = true
+    if (signedUserPhoneNumber) {
+      let isValidSig = false
+      try {
+        isValidSig = await verifySignedUserPhoneNumberSignature(
+          account,
+          userPhoneNumber,
+          signedUserPhoneNumber,
+          logger
+        )
+      } catch (error) {
+        if (error.message === ErrorMessage.CONTRACT_GET_FAILURE) {
+          logger.warn(
+            'Failed to retrieve DEK to verify signedUserPhoneNumber. Signature is assumed valid and request is not recorded.'
+          )
+          isValidSig = true
+          shouldStoreRequest = false
+        } else {
+          throw error
+        }
+      }
+      if (!isValidSig) {
+        respondWithError(response, 403, WarningMessage.INVALID_USER_PHONE_NUMBER_SIGNATURE, logger)
+        return
+      }
+
+      // Hash signed number one more time for good measure
+      signedUserPhoneNumberHash = crypto
+        .createHash('sha256')
+        .update(signedUserPhoneNumber)
+        .digest('base64')
+    }
+
     if (await getDidMatchmaking(account, logger)) {
-      respondWithError(response, 403, WarningMessage.DUPLICATE_REQUEST_TO_MATCHMAKE, logger)
-      return
+      logger.info({ account }, 'Account has performed matchmaking before.')
+      if (!signedUserPhoneNumber) {
+        // If the account has performed matchmaking before and does not provide their signed phone number in the request,
+        // we return an error bc they could be querying matches for a new number that isn't theirs.
+        logger.info(
+          { account },
+          'Blocking account from requerying matches without a phone number signature.'
+        )
+        respondWithError(response, 403, WarningMessage.DUPLICATE_REQUEST_TO_MATCHMAKE, logger)
+        return
+      } else {
+        // Account has performed matchmaking before and has provided a phone number dek signature
+        const signedUserPhoneNumberRecord = await getAccountSignedUserPhoneNumberRecord(
+          account,
+          logger
+        )
+        if (!signedUserPhoneNumberRecord) {
```

Comment:
TODO: this should check for when the record doesn't show up due to a db error and skip storing the request in that case